graviti/schema/pyarrow.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+"""The PyArrow extension types to represent PortexType."""
+
+import json
+from typing import Any, Dict, Optional, Type, TypeVar
+
+import pyarrow as pa
+from pyarrow.lib import DataType  # pylint: disable=no-name-in-module
+
+# https://pylint.pycqa.org/en/latest/technical_reference/c_extensions.html
+
+
+class BuiltinExtension(  # pylint: disable=too-few-public-methods
+    pa.ExtensionType  # type: ignore[misc]
+):
+    """This class defines the PyArrow representation of PortexBuiltinType builtin types.
+
+    Arguments:
+        portex_type_name: The PortexBuiltinType name.
+        nullable: Whether the PortexBuiltinType is nullable.
+        storage_type: The corresponding PyArrow type.
+        kwargs: Other keyword arguments to init the PortexBuiltinType.
+
+    """
+
+    _T = TypeVar("_T", bound="BuiltinExtension")
+
+    def __init__(
+        self,
+        portex_type_name: str,
+        storage_type: DataType,
+        *,
+        nullable: bool = False,
+        **kwargs: Any,
+    ):
+        self._portex_type_name = portex_type_name
+        self._nullable = nullable
+        self._kwargs = kwargs
+        pa.ExtensionType.__init__(self, storage_type, f"builtin.{portex_type_name}")
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return json.dumps(self._dumps()).encode("utf-8")
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls: Type[_T], storage_type: Any, serialized: bytes) -> _T:
+        kwargs = json.loads(serialized.decode("utf-8"))
+        return cls(**kwargs, storage_type=storage_type)
+
+    def _dumps(self) -> Dict[str, Any]:
+        content = {"portex_type_name": self._portex_type_name, "nullable": self._nullable}
+        if self._kwargs:
+            content.update(self._kwargs)
+        return content
+
+
+class ExternalExtension(BuiltinExtension):  # pylint: disable=too-few-public-methods
```

Comment:
External is a subclass of Builtin? Will it be better if External and Builtin are subclasses of a base class?